### PR TITLE
adds eval mode to stateful module

### PR DESCRIPTION
**Pull request by RickSanchezStoic**
Adds a recursive eval function that iterates through modules to set the eval mode to the specified mode if `training` is an attribute of the said child modules or the module itself. Fixed BatchNoorm2D to maintain the internal state of training.

### CODE REVIEW

File: ivy/stateful/module.py [hunk 1]
```diff
@@ -817,6 +817,16 @@ def register_buffer(self, var_name, value):
         """Set the buffer at any place within the class."""
         self._set_buffers({var_name: value})
 
+    def eval(self, mode: bool = False):
```
Comment (vedpatwardhan) [inline ivy/stateful/module.py:820]: Actually `eval` doesn't need to receive a `mode` argument, basically `model.train()` would set `training` to `True` and `model.eval()` would set `training` to `False` as used in `torch.nn.Module`. Also `self.training` should be `True` by default, so we should be setting that in the `__init__` itself

File: ivy/stateful/module.py [hunk 2]
```diff
@@ -817,6 +817,16 @@ def register_buffer(self, var_name, value):
...
+        # disables training mode for child modules
+        # if it exists
+        if hasattr(self, "training"):
+            self.training = mode
+        for module in self.v:
```
Comment (vedpatwardhan) [inline ivy/stateful/module.py:825]: `self.v` is a `ivy.Container` with `ivy.Array` leaves right? Not sure why it would have an `ivy.Module` instance. Did you intend to use `self.__dict__` instead?

File: ivy/stateful/norms.py [hunk 3]
```diff
@@ -151,6 +152,7 @@ def __init__(
         self._bias_init = Zeros()
         self._running_mean_init = Zeros()
         self._running_var_init = Ones()
+        self.training = training
```
Comment (vedpatwardhan) [inline ivy/stateful/norms.py:155]: Actually we wouldn't need to set a `training` variable for `LayerNorm` or `Dropout` explicitly, all this should be handled directly through whether `self.training` was `True` for the parent module (which could of course also be this same module)

File: ivy/stateful/norms.py [hunk 4]
```diff
@@ -175,7 +177,7 @@ def _create_variables(self, device, dtype=None):
     def _forward(
         self,
         inputs,
-        training: bool = False,
+        training: bool = None,
```
Comment (vedpatwardhan) [inline ivy/stateful/norms.py:180]: let's remove this argument, I think we might just introduce confusion by adding an additional argument where `model.train()` and `model.eval()` should really be the only way to control the training mode passed. So provided we ensure that, only passing `self.training` on line [206](https://github.com/unifyai/ivy/pull/22469/files#diff-08ded1880c50b8c039f86533aab4b2b05be766fe26e9c7edc8d4655f4acb6149R206) without any additional handling should work

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Hey @RickSanchezStoic, thanks for looking into this, just requested a few changes 😄

File: ivy/stateful/module.py (hunk 2 above)
Comment (RickSanchezStoic) [reply ivy/stateful/module.py:825]: I am iterating the keys, which are going to be Module instances. Iterating over `self.__dict__` will be suboptimal, as we need to just focus on the Module instances that are contributing to weight updates.

Comment (vedpatwardhan) [reply ivy/stateful/module.py:825]: yeah you're right, thanks for clarifying!

File: ivy/stateful/norms.py [hunk 5]
```diff
@@ -192,6 +194,8 @@ def _forward(
         ret
             The outputs following the batch normalization operation.
         """
```
Comment (mattbarrett98) [inline ivy/stateful/norms.py:196]: I agree with ved, I'm not seeing why this logic is necessary. training mode is now dictated by `self.training` so surely no need for `training` to also be an argument to the forward, and we just pass `self.training` to the functional batch norm directly?

**Comment by mattbarrett98** [comment]
+ maybe add a test

File: ivy/stateful/module.py (hunk 1 above)
Comment (vedpatwardhan) [reply ivy/stateful/module.py:821]: There's a minor bug with this logic where if a user passes `model.eval(True)`, it would set `training=True`

File: ivy/stateful/module.py [hunk 6]
```diff
@@ -817,6 +818,18 @@ def register_buffer(self, var_name, value):
...
+        # disables training mode for child modules
+        self.training = mode
+        for module in self.v:
+            module = getattr(self, module, None)
+            if isinstance(module, ivy.Module):
+                module.eval()
+
+    def train(self, mode: bool = True):
+        # enables/disables training mode
+        self.eval(mode=mode)
```
Comment (vedpatwardhan) [inline ivy/stateful/module.py:831]: this should actually be `self.eval(mode = not mode)`

File: ivy/stateful/norms.py (hunk 4 above)
Comment (vedpatwardhan) [reply ivy/stateful/norms.py:180]: but how would this be different from doing a `model.train()` or `model.eval()` on a BatchNorm module. If we consider the `torch` design for e.g., [`torch.nn.functional.batch_norm`](https://pytorch.org/docs/stable/generated/torch.nn.functional.batch_norm.html) has a `training` parameter, but [`torch.nn.BatchNorm2d`](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html#torch.nn.BatchNorm2d) doesn't have such a parameter, and it doesn't need it either given that there's `model.train()` and `model.eval()` 😅
Secondly it becomes hard to debug if this `training` parameter overwrites `self.training` because we already have `model.train()` and `model.eval()` to do that.
To your point of being able to control this to the module, we could accept `training` as an argument to the `ivy.Module.__init__` which would be `True` by default, and then once a `ivy.Module` is created, the only way to modify it would be by calling `model.train()` or `model.eval()` (for the parent module or any child module) 😄

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]

File: ivy/stateful/norms.py (hunk 4 above)
Comment (RickSanchezStoic) [reply ivy/stateful/norms.py:180]: In that case, we would need to get rid of the argument from the _forward method of this implementation.

File: ivy/stateful/norms.py (hunk 5 above)
Comment (RickSanchezStoic) [reply ivy/stateful/norms.py:196]: agreed, we can get rid of the argument from forward entirely.
